packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -579,48 +579,82 @@ impl<'m> RunnerImpl<'m> {
-            },
-            DataType::UInt16 => unsafe {
-                let values = slice::from_raw_parts(data.buffer_ptrs[0] as *const u16, target_len);
+            }
+            DataType::UInt16 => {
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `u16`, and
+                //   `target_len` is carefully chosen
+                let values =
+                    unsafe { slice::from_raw_parts(data.buffer_ptrs[0] as *const u16, target_len) };
+
                 builder = builder.add_buffer(Buffer::from_slice_ref(&values));
-            },
-            DataType::UInt32 => unsafe {
-                let values = slice::from_raw_parts(data.buffer_ptrs[0] as *const u32, target_len);
+            }
+            DataType::UInt32 => {
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `u32`, and
+                //   `target_len` is carefully chosen
+                let values =
+                    unsafe { slice::from_raw_parts(data.buffer_ptrs[0] as *const u32, target_len) };
+
                 builder = builder.add_buffer(Buffer::from_slice_ref(&values));
-            },
-            DataType::Float64 => unsafe {
-                let values = slice::from_raw_parts(data.buffer_ptrs[0] as *const f64, target_len);
+            }
+            DataType::Float64 => {
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `f64`, and
+                //   `target_len` is carefully chosen
+                let values =
+                    unsafe { slice::from_raw_parts(data.buffer_ptrs[0] as *const f64, target_len) };
+
                 builder = builder.add_buffer(Buffer::from_slice_ref(&values));
-            },
-            DataType::Utf8 => unsafe {
+            }
+            DataType::Utf8 => {
+                // Utf8 is stored in two buffers:
+                //   [0]: The offset buffer (i32)
+                //   [1]: The value buffer (u8)
+
                 // TODO: Use `data.len` or target_len?
-                //       (In practice, target_len has worked for a long time,
-                //       though that's not an ideal reason to use it. Maybe
-                //       `data.len` would also work.)
-                let offsets =
-                    slice::from_raw_parts(data.buffer_ptrs[0] as *const i32, target_len + 1);
+                //   (In practice, target_len has worked for a long time, though that's not an ideal
+                //   reason to use it. Maybe `data.len` would also work.)
+                // The offsets are in the first buffer. For each value in the second buffer, we have
+                // a start offset and an end offset. The start offset is equal to the end offset of
+                // the previous value, thus we need `num_values + 1` offset values.
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow as offsets, the type is `i32`,
```

Comment:
Can we reorder these comments and add a line-break between the bits, like the TODO and the actual comment and also the SAFETY point

Reply:
I added a newline between the comment and the SAFETY section. The Todo was removed as we don't want to use `data.len` anymore.